cupy/cuda/nccl.pyx

```diff
@@ -145,6 +150,46 @@ cdef class NcclCommunicator:
     def __dealloc__(self):
         self.destroy()
 
+    @staticmethod
+    def initAll(int ndev, list devlist=None):
+        """ Initialize NCCL communicator in a single process.
+
+        Args:
+            ndev (int): Number of GPUs to be used.
+            devlist (None or list): A list of integers that label the GPUs to be
+                used. The default is None, meaning that the first ndev GPUs will
+                be chosen.
+
+        Returns:
+            NcclCommunicator: An NcclCommunicator instance.
+
+        .. note::
+            This method is to be used to create an NCCL communicator in a single
+            process like this:
+
+            .. code-block:: python
+
+                from cupy.cuda import nccl
+                # Use GPU #0, #2, and #3
+                comm = nccl.NcclCommunicator.initAll(3, [0, 2, 3])
+
+            In a multi-process setup, use the default initializer instead.
+        """
+        cdef array.array devices
+        cdef int * devices_ptr
+        # Call to __new__ bypasses __init__ constructor
+        cdef NcclCommunicator NcclComm = \
+            NcclCommunicator.__new__(NcclCommunicator)
+        if devlist is not None:
+            assert len(devlist) == ndev
+            devices = array.array('i', devlist)
```

Comment:
I am not sure if CuPy has a convention for handling `list` <-> `int *`, so I used Cython's `array.array` feature.

Reply:
Switched to `vector.vector` throughout.